### ISSUE TITLE
Bad performance of targeted versions of AutoPGD and PGD attacks 

### ISSUE THREAD

**Issue by bezirganyan** [Trusted-AI/adversarial-robustness-toolbox#840]
I've tested several targeted adversarial attacks (FGSM, PGD, AutoPGD and Fast Wasserstein (not from this framework)) in the same setup: Here you can see the scores for different epsilon values and l_inf norm :
![Screenshot_20210114_144442](https://user-images.githubusercontent.com/16824193/104598604-0df21d00-5677-11eb-8cbf-9a8a38264fbc.png)

and here, the best score for an epsilon in [0, 0.5] interval and specified norm. 
![Screenshot_20210114_145705](https://user-images.githubusercontent.com/16824193/104600047-c9678100-5678-11eb-96dc-8d827d500f82.png)

From the results we notice two strange things:
1. AutoPGD has really bad performance although claiming to have state of the art results
2. PGD isn't affected by the choice of epsilon 

You can find the code for comparison [here](https://github.com/bezirganyan/adversarial_arena)

**To Reproduce**

Run the targeted PGD and targeted AutoPGD attacks on CIFAR 10 with different epsilon values 

**Expected behavior**
* PGD shall change the score depending on epsilon
* AutoPGD shall perform better, according to the paper

**System information:**
 - Linux
 - Python 3.8.5
 - ART version: 1.5.0
 - PyTorch 1.7.0

**Comment by beat-buesser** [comment]
Hi @bezirganyan Thank you very much for using ART!

Thank you for providing the code of your experiments, to get a better understanding I have a few questions: Is your model trained on normalised or un-normalised data? What is the accuracy on benign test data as calculated in line [L140](https://github.com/bezirganyan/adversarial_arena/blob/1b350f123999833f78bb0319591625dd8be57802/generate.py#L140)

**Comment by bezirganyan** [comment]
Hi @beat-buesser, Thanks a lot for the reply. 

The results are from model trained on an un-normalized data. I also have the one on normalized data, but, I don't have the results of that training anymore. 

The accuracy on benign test data is 92.4%.

**Comment by beat-buesser** [comment]
Ok, that sounds good.

I looked closer at your code and noticed [L112 and L113](https://github.com/bezirganyan/adversarial_arena/blob/1b350f123999833f78bb0319591625dd8be57802/generate.py#L112) overwriting attributes. I'm not sure if this will solve the problem entirely but it might explain why changing `eps` does not change the mis-classifications of PGD. `ProjectedGradientDescent` is a wrapper around the framework-specific implementations and the changed attributes are not passed to the actual internal attack. You can either set the attributes with `ProjectedGradientDescent.set_params` or use `ProjectedGradientDescentPyTorch` directly to automate setting the parameters.

Would you be able to share the weights for one of the models to run your script?

**Comment by bezirganyan** [comment]
Oh, I see the problem regarding the PGD. Will try again with set_params. 
The problem with AutoPGD is independent from that though, right? It responds to epsilon change, but behaves really poorly. 

[Here](https://drive.google.com/file/d/1ok4xVg3dvaMpD-lWZSgLKsxYbv0oZBAv/view?usp=sharing) you can find the trained model for reproducing the results. 

Also I just pushed a small bug fix, so please do a git pull before running the code. The bug fix will not anyhow affect the strange behavior, but will prevent from results overwriting each other.

**Comment by beat-buesser** [comment]
@bezirganyan Thank you for sharing the model and the updated code, I'll try it out later today.

Yes, AutoPGD should not be affected by the attributes change, although `set_params` is also here always the safer approach in case of internal changes in the future. The last time I tested AutoPGD it was slightly better in accuracy than PGD while achieving much smaller perturbations for the test model we used, but we made some changes to AutoPGD recently, therefore I'm very interested to learn how AutoPGD performs in your experiments and to see if something needs to be improved.

**Comment by beat-buesser** [comment]
@bezirganyan I was able to run your code. So far, one thing I noticed is  that your code is using `art.utils.to_categorical` in two calls with the default argument `nb_classes=None`. This leads to unexpected target labels as `to_categorical` tries to derive the total number of categories from the provided labels which doesn't work for targeted labels, in this case `to_categorical` needs to be provided as `to_categorical(target_classes, nb_classes=10)` in lines 71 and 76.

**Comment by bezirganyan** [comment]
@beat-buesser thanks. I rerun the code with your suggestions. From the logs it seems that the problem with PGD is resolved, but AutoPGD still performs  poorly. Will keep you updated as soon as I summarize the results.

**Comment by bezirganyan** [comment]
@beat-buesser Here is the summarized version of the updated analysis. 
![image](https://user-images.githubusercontent.com/16824193/105375860-64c49d00-5c09-11eb-8d6b-8353c606b6ef.png)

As you can see the PGD attack now works as expected, while AutoPGD continues to behave poorly, while I believe it had to perform at least as good as the vanilla PGD attack.

**Comment by beat-buesser** [comment]
Hi @bezirganyan Thank you very much for the updated experiment code and results.

I looked further into it and I think I have discovered two issues with targeted `AutoProjectedGradientDescent` that most likely cause your observations:

- in the experiments on this (I think) undefended model, the loss value seems to decreases very fast to often numerical 0 already before the first check on having the step size, the loss stays there but it suddenly changes to larger values which I think is caused by amplified random numerical noise in the gradients based on a loss value close to 0. I fixed this with a break condition.

- after fixing the first issue I noticed that the criterion for counting if a step got into the right direction should depend on if the attack is targeted or untargeted, currently it is defined for untargeted attacks where the loss criteria is expected to only increase. I modified it to expected a decreasing value for the loss criteria in targeted attacks.

I have pushed the updates to https://github.com/Trusted-AI/adversarial-robustness-toolbox/pull/860

Could you please test it and let us know how it performs in your experiments?

**Comment by bezirganyan** [comment]
Hello @beat-buesser. Thanks a lot for the fix. I rerun the AutoPGD and now it performs similar to PGD. 
Here you can see the results

![image](https://user-images.githubusercontent.com/16824193/105770108-0e37c580-5f5f-11eb-90e3-e148b9066415.png)

Now, I suppose it does not perform better, since the model is not anyhow protected. From your own experiments on protected models, does it perform much better than the PGD attack? 

Furthermore, when I tried to run the code with the updated version, I got the error"

```
  File "generate.py", line 71, in attack_model
    x_test_adv = attacker.generate(x=x_test.cpu(), y=to_categorical(target_classes.cpu(), nb_classes=10))
  File "/usr/local/share/anaconda3/lib/python3.7/site-packages/art/attacks/attack.py", line 74, in replacement_function
    return fdict[func_name](self, *args, **kwargs)
  File "/usr/local/share/anaconda3/lib/python3.7/site-packages/art/attacks/evasion/auto_projected_gradient_descent.py", line 474, in generate
    f_0 = self.estimator.loss(x=x_k, y=y_batch, reduction="mean")
  File "/usr/local/share/anaconda3/lib/python3.7/site-packages/art/estimators/classification/pytorch.py", line 508, in loss
    return loss.detach().numpy()
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```
Which I fixed by changing `return loss.detach().numpy()` to `return loss.cpu().detach().numpy()` in `art/estimators/classification/pytorch.py` file.

**Comment by beat-buesser** [comment]
Hi @bezirganyan That's great, thank you very much for the updated results and also for reporting the issue in the first place, the solution of it has already been helpful for other users of ART!

In my experiments on untargeted attacks AutoPGD achieved slightly better attack success in terms of accuracy but with a smaller adversarial perturbations than PGD. An additional interesting visualisation in your experiments could therefore be the mean adversarial perturbation per image and its standard deviation for AutoPGD and PGD as function of `eps`. Also, AutoPGD and PGD have different default values for the number of iteration and number of random restarts which you could consider before comparing them directly, I think the best would be to optimise these parameters until convergence of the attack success before comparison. You also could then repeat the same experiment with an adversarially trained model e.g. https://github.com/MadryLab/robustness

Thank you for the error message, we have opened a pull request with a fix in #862
